### PR TITLE
feat: add telegram messageThreadID

### LINKED ISSUE #6997: AlertManagerConfig - Telegram messageThreadID
### Component(s)

AlertManagerConfig

### What is missing? Please describe.

I currently configure alertmanager to send messages to Telegram Group. We use Topics in our Telegram Group. 
Currently it is not possible to send messages to topics because I cant define the message thread id. 
I saw in the alertmanager repo that this is already implemented, but not in the alertmanagerconfig crd. --> https://github.com/prometheus/alertmanager/blob/main/config/notifiers.go#L779

Will this somehow be implemented?

### Describe alternatives you've considered.

_No response_

## Environment
Kubernetes Version:
Prometheus-Operator Version:

### PR DESCRIPTION
**Pull request by dcardellino**
This adds the functionality to configure the telegram message thread id.

## Description

Fixes #6997

This adds the functionality to configure the telegram message thread id.

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Verification

Please check the [Prometheus-Operator testing guidelines](../TESTING.md) for recommendations about automated tests.

## Changelog entry

_Please put a one-line changelog entry below. This will be copied to the changelog file during the release process._

```release-note
feat: add messageThreadID in TelegramConfig
```

### CODE REVIEW

**Comment by slashpai** [comment]
Can you add unit test as well?

**Comment by slashpai** [comment]
Here is an example https://github.com/prometheus-operator/prometheus-operator/pull/5882 for adding field to AlertmanagerConfig if that helps :)

**Comment by dcardellino** [comment]
@slashpai So you mean like `versionTelegramBotTokenFileAllowed`?

messageThreadID is an optional field, is there really a need to test this? If you really want to I will implement it :)

**Comment by slashpai** [comment]
> @slashpai So you mean like `versionTelegramBotTokenFileAllowed`?
> 
> `messageThreadID` is an optional field, is there really a need to test this? If you really want to I will implement it :)

You should have a unit test regardless field is optional or not to verify its added correctly. In example PR we check for version also from which field was supported in Alertmanager

**Comment by dcardellino** [comment]
@slashpai 
Added some basic unit tests for the field in the alertmanager configuration.

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 1]
```diff
@@ -912,6 +912,9 @@ type TelegramConfig struct {
 	// The Telegram chat ID.
 	// +required
 	ChatID int64 `json:"chatID,omitempty"`
+	// The Telegram Group Topic ID
+	// +optional
+	MessageThreadID int64 `json:"messageThreadID,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:917]:
```suggestion
	MessageThreadID *int64 `json:"messageThreadID,omitempty"`
```

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go [hunk 2]
```diff
@@ -903,6 +903,9 @@ type TelegramConfig struct {
 	// The Telegram chat ID.
 	// +required
 	ChatID int64 `json:"chatID,omitempty"`
+	// The Telegram Group Topic ID
+	// +optional
+	MessageThreadID int64 `json:"messageThreadID,omitempty"`
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:908]:
```suggestion
	MessageThreadID *int64 `json:"messageThreadID,omitempty"`
```

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go [hunk 3]
```diff
@@ -912,6 +912,9 @@ type TelegramConfig struct {
 	// The Telegram chat ID.
 	// +required
 	ChatID int64 `json:"chatID,omitempty"`
+	// The Telegram Group Topic ID
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:915]:
```suggestion
	// The Telegram Group Topic ID.
	// It requires Alertmanager >= 0.26.0.
```

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go [hunk 4]
```diff
@@ -903,6 +903,9 @@ type TelegramConfig struct {
 	// The Telegram chat ID.
 	// +required
 	ChatID int64 `json:"chatID,omitempty"`
+	// The Telegram Group Topic ID
```
Comment (simonpasquier) [inline pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:906]:
```suggestion
	// The Telegram Group Topic ID.
	// It requires Alertmanager >= 0.26.0.
```

File: pkg/alertmanager/amcfg.go [hunk 5]
```diff
@@ -1210,6 +1210,7 @@ func (cb *configBuilder) convertTelegramConfig(ctx context.Context, in monitorin
 		VSendResolved:        in.SendResolved,
 		APIUrl:               in.APIURL,
 		ChatID:               in.ChatID,
+		MessageThreadID:      in.MessageThreadID,
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:1213]: you'll need also to update the *telegramConfig.sanitize() function to drop the message thread id field if version < 0.26.0.

File: pkg/alertmanager/amcfg_test.go [hunk 6]
```diff
@@ -2939,6 +2942,57 @@ func TestSanitizeConfig(t *testing.T) {
...
+							{
+								ChatID:          12345,
+								MessageThreadID: 123,
+								BotTokenFile:    "/test",
+							},
+						},
+					},
+				},
+			},
+			expectErr: true,
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg_test.go:2994]: the returned error isn't the one expected (it says `missing mandatory field botToken or botTokenFile`). You need to set `botToken` instead.

**Review by simonpasquier — COMMENTED** [review]
Thanks for contributing! Just a few nits :)

File: pkg/alertmanager/amcfg_test.go (hunk 6 above)
Comment (dcardellino) [reply pkg/alertmanager/amcfg_test.go:2994]: @simonpasquier Can you just support me here? The changes above broke my tests..?

**Comment by dcardellino** [comment]
@simonpasquier So should be fixed all now. Ready to be merged

**Comment by dcardellino** [comment]
Any update here? I really would love to use this function as I need it for a project.

**Review by slashpai — APPROVED** [review]
lgtm, but I would let @simonpasquier merge :)
If merged this would be available in next release which is on 30th Oct :)

**Comment by dcardellino** [comment]
Awesome 👍 happy to hear 😊

File: pkg/apis/monitoring/v1beta1/alertmanager_config_types.go (hunk 2 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1beta1/alertmanager_config_types.go:909]: We still need this change

File: pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go (hunk 1 above)
Comment (simonpasquier) [reply pkg/apis/monitoring/v1alpha1/alertmanager_config_types.go:918]: We need this change.

**Review by simonpasquier — CHANGES_REQUESTED** [review]

**Comment by dcardellino** [comment]
@simonpasquier I committed your requested changes. But now tests are failing. Can you check?

**Comment by slashpai** [comment]
You need to run generate after updating types

` make --always-make format generate && git diff --exit-code`

**Comment by dcardellino** [comment]
@slashpai 
I ran this locally but I get this:

```
# github.com/prometheus-operator/prometheus-operator/pkg/alertmanager
pkg/alertmanager/amcfg.go:1202:25: cannot use in.MessageThreadID (variable of type *int64) as int64 value in struct literal
make: *** [operator] Error 1
```

File: pkg/alertmanager/types.go [hunk 7]
```diff
@@ -353,6 +353,7 @@ type telegramConfig struct {
 	BotToken             string            `yaml:"bot_token,omitempty" json:"bot_token,omitempty"`
 	BotTokenFile         string            `yaml:"bot_token_file,omitempty" json:"bot_token_file,omitempty"`
 	ChatID               int64             `yaml:"chat_id,omitempty" json:"chat_id,omitempty"`
+	MessageThreadID      int64             `yaml:"message_thread_id,omitempty" json:"message_thread_id,omitempty"`
```
Comment (Xu-Wentao) [inline pkg/alertmanager/types.go:356]:
```suggestion
	MessageThreadID      *int64             `yaml:"message_thread_id,omitempty" json:"message_thread_id,omitempty"`
```

**Comment by dcardellino** [comment]
@slashpai @simonpasquier 

I do not know what i am doing wrong but I get:
```
pkg/alertmanager/amcfg.go:2262:27: invalid operation: tc.MessageThreadID != 0 (mismatched types *int64 and untyped int)
pkg/alertmanager/amcfg.go:2265:24: cannot use 0 (untyped int constant) as *int64 value in assignment
```

Help would be much appreciated..

File: pkg/alertmanager/amcfg.go [hunk 8]
```diff
@@ -2258,6 +2259,12 @@ func (tc *telegramConfig) sanitize(amVersion semver.Version, logger *slog.Logger
 		tc.BotTokenFile = ""
 	}
 
+	if tc.MessageThreadID != 0 && lessThanV0_26 {
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:2262]:
```suggestion
	if tc.MessageThreadID != nil && lessThanV0_26 {
```

File: pkg/alertmanager/amcfg.go [hunk 9]
```diff
@@ -2258,6 +2259,12 @@ func (tc *telegramConfig) sanitize(amVersion semver.Version, logger *slog.Logger
...
+		msg := "'message_thread_id' supported in Alertmanager >= 0.26.0 only - dropping field from provided config"
+		logger.Warn(msg, "current_version", amVersion.String())
+		tc.MessageThreadID = 0
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:2265]:
```suggestion
		tc.MessageThreadID = nil
```

File: pkg/alertmanager/types.go [hunk 10]
```diff
@@ -353,6 +353,7 @@ type telegramConfig struct {
 	BotToken             string            `yaml:"bot_token,omitempty" json:"bot_token,omitempty"`
 	BotTokenFile         string            `yaml:"bot_token_file,omitempty" json:"bot_token_file,omitempty"`
 	ChatID               int64             `yaml:"chat_id,omitempty" json:"chat_id,omitempty"`
+	MessageThreadID      *int64            `yaml:"message_thread_id,omitempty" json:"message_thread_id,omitempty"`
```
Comment (simonpasquier) [inline pkg/alertmanager/types.go:356]: here we want to mimic the native Alertmanager config struct.
```suggestion
	MessageThreadID      int            `yaml:"message_thread_id,omitempty" json:"message_thread_id,omitempty"`
```

File: pkg/alertmanager/amcfg.go [hunk 11]
```diff
@@ -1199,6 +1199,7 @@ func (cb *configBuilder) convertTelegramConfig(ctx context.Context, in monitorin
 		VSendResolved:        in.SendResolved,
 		APIUrl:               in.APIURL,
 		ChatID:               in.ChatID,
+		MessageThreadID:      in.MessageThreadID,
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:1202]:
```suggestion
```
and you need this after

```golang

       if in.MessageThreadID != nil {
               out.MessageThreadID = int(*in.MessageThreadID)
       }
````

**Comment by dcardellino** [comment]
@simonpasquier Okay comitted the suggestions:

```
invalid operation: tc.MessageThreadID != nil (mismatched types int and untyped nil)
Error: pkg/alertmanager/amcfg.go:2268:24: cannot use nil as int value in assignment
```

**Comment by mviswanathsai** [comment]
Quick question, why are there two (actually, three) files which define types for AlertmanagerConfig? 
I see 
1. `pkg/api/monitoring/v1beta1/alertmanager_config_types.go`
2. `pkg/api/monitoring/v1alpha1/alertmanager_config_types.go`
3. and then another in `alertmanager/types.go`

I get the existence of the first two, but why do we need the third one? More specifically, why are we repeating the same definitions again?
I see different parts of the code uses different types (either defined in `types.go` or the ones in `/api/monitoring`) and it's all very confusing.

Can anyone explain? @simonpasquier @slashpai

**Comment by slashpai** [comment]
@mviswanathsai

I agree that its a little confusing :)

We have 2 API versions for the AlertmanagerConfig CRD: v1alpha1 and v1beta1, former being the default stored version.
Hence adding adding both is needed now. 

The reason for adding in `alertmanager/types.go`  is we have two ways to define alertmanager configs one as a yaml config hence this file exists. Another way is through AlertmanagerConfig CRD.

https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/alerting.md explains about this. Also a documentation on this is in progress https://github.com/prometheus-operator/prometheus-operator/pull/6897

@simonpasquier is most experienced with alertmanager so he could possibly add more inputs if I missed :)

**Comment by mviswanathsai** [comment]
Right, so the `alertmanager/types.go` is to marshal/un-marshal the secret stored in yaml format. Would it also follow similar conventions? for e.g, pointers for optional fields, etc. 
I see that the linter error is related to having `messageThreadID` defined as type `int` and we are checking for `!= nil` which is invalid. Either we keep it and check for defaults, or change it to a pointer? 

What's more is, there are multiple sanity checks that happen and the pipeline is different for the yaml and AlertmanagerConfig methods of defining the config. I might be wrong, but I think the PR does not yet cover both the pipelines?

**Comment by slashpai** [comment]
Since that's not a CRD we don't follow same rules AFAIK

File: pkg/alertmanager/amcfg.go [hunk 12]
```diff
@@ -2258,6 +2262,12 @@ func (tc *telegramConfig) sanitize(amVersion semver.Version, logger *slog.Logger
 		tc.BotTokenFile = ""
 	}
 
+	if tc.MessageThreadID != nil && lessThanV0_26 {
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:2265]:
```suggestion
	if tc.MessageThreadID > 0 && lessThanV0_26 {
```

File: pkg/alertmanager/amcfg.go [hunk 13]
```diff
@@ -2258,6 +2262,12 @@ func (tc *telegramConfig) sanitize(amVersion semver.Version, logger *slog.Logger
...
+		msg := "'message_thread_id' supported in Alertmanager >= 0.26.0 only - dropping field from provided config"
+		logger.Warn(msg, "current_version", amVersion.String())
+		tc.MessageThreadID = nil
```
Comment (simonpasquier) [inline pkg/alertmanager/amcfg.go:2268]:
```suggestion
		tc.MessageThreadID = 0
```

**Comment by dcardellino** [comment]
So there you go. Thanks @simonpasquier for your support!

**Review by simonpasquier — APPROVED** [review]
thanks!

**Pull request merged**